pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaGroupPartitionSelector.java

```diff
@@ -180,14 +266,104 @@ public void selectInstances(Map<Integer, List<InstanceConfig>> poolToInstanceCon
+    // Initialize the list with empty positions to fill.
+    List<String> instancesToSelect = new ArrayList<>(numInstancesToSelect);
+    for (int i = 0; i < numInstancesToSelect; i++) {
+      instancesToSelect.add(null);
+    }
+    Deque<String> newlyAddedInstances = new LinkedList<>();
+
+    // Find out the existing instances that are still alive.
+    Set<String> existingInstancesStillAlive = new HashSet<>();
+    for (String existingInstance : existingInstances) {
+      if (candidateInstances.contains(existingInstance)) {
+        existingInstancesStillAlive.add(existingInstance);
+      }
+    }
+
+    // Find out the newly added instances.
+    for (String candidateInstance : candidateInstances) {
+      if (!existingInstancesStillAlive.contains(candidateInstance)) {
+        newlyAddedInstances.add(candidateInstance);
+      }
+    }
+
+    int numExistingInstances = existingInstances.size();
+    for (int i = 0; i < numInstancesToSelect; i++) {
+      String existingInstance = i < numExistingInstances ? existingInstances.get(i) : null;
+      String selectedInstance;
+      if (existingInstance != null && candidateInstances.contains(existingInstance)) {
+        selectedInstance = existingInstance;
+        existingInstancesStillAlive.remove(selectedInstance);
+      } else {
+        selectedInstance = newlyAddedInstances.poll();
+      }
+      instancesToSelect.set(i, selectedInstance);
+      // If it's an existing alive instance, or it's for a new replica group, add the new instance to the tail,
+      // so that it won't be firstly chosen for the next partition.
+      // For newly added instances to fill the existing replica group, the sequence cannot change;
+      // otherwise there is no guarantee that same vacant position will be filled with the same new instance.
+      // The 'selectedInstance' object can still be null if there is no new instances from the candidate list.
+      if (selectedInstance != null && (i < numExistingInstances || existingInstances.isEmpty())) {
+        candidateInstances.remove(selectedInstance);
+        candidateInstances.add(selectedInstance);
+      }
+    }
+
+    // If there are still some vacant positions in the instance list,
+    // try to fill with instances which are either left over or newly added.
+    for (int i = 0; i < instancesToSelect.size(); i++) {
+      if (instancesToSelect.get(i) == null) {
+        if (!existingInstancesStillAlive.isEmpty()) {
+          Iterator<String> iterator = existingInstancesStillAlive.iterator();
+          String existingInstanceLeftOver = iterator.next();
+          instancesToSelect.set(i, existingInstanceLeftOver);
+          iterator.remove();
+        } else if (!newlyAddedInstances.isEmpty()) {
+          // pick a new instance to fill its vacant position.
+          String newInstance = newlyAddedInstances.pollFirst();
+          instancesToSelect.set(i, newInstance);
+        }
+      }
+    }
```

Comment:
This loop shouldn't be needed

Reply:
This is still needed for the scenarios when some of the instances got removed, then the instances from the tail will fill the vacant positions. 
E.g. supposed the existing instance assignment is [i1, i2, i3, i4, i5], and if i3 is removed/decommissioned, the very last instance (i.e. i5) should fill i3's vacant position.